analyzers/src/SonarAnalyzer.CSharp/Rules/SillyMathematicalComparison.cs

```diff
@@ -0,0 +1,116 @@
+
+        private static readonly DiagnosticDescriptor MathComparisonRule = DescriptorFactory.Create(DiagnosticId, MathComparisonMessage);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(MathComparisonRule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterNodeAction(
+                CheckBinary,
+                SyntaxKind.GreaterThanExpression,
+                SyntaxKind.GreaterThanOrEqualExpression,
+                SyntaxKind.LessThanExpression,
+                SyntaxKind.LessThanOrEqualExpression,
+                SyntaxKind.EqualsExpression,
+                SyntaxKind.NotEqualsExpression);
+
+        private static void CheckBinary(SonarSyntaxNodeReportingContext context)
+        {
+            if (TryGetConstantValue(context, (BinaryExpressionSyntax)context.Node, out var constant, out var other)
+               && context.SemanticModel.GetTypeInfo(other).Type is { } typeSymbolOfOther
+               && TryGetRange(typeSymbolOfOther, out var min, out var max)
+               && (constant < min || constant > max))
+            {
+                var typeName = context.SemanticModel.GetTypeInfo(other).Type.ToMinimalDisplayString(context.SemanticModel, other.GetLocation().SourceSpan.Start);
+                context.ReportIssue(Diagnostic.Create(MathComparisonRule, other.Parent.GetLocation(), typeName));
+            }
+        }
+
+        private static bool TryGetConstantValue(SonarSyntaxNodeReportingContext context, BinaryExpressionSyntax binary, out double constant, out SyntaxNode other)
+        {
+            constant = default;
+            other = default;
+            var maybeLeft = context.SemanticModel.GetConstantValue(binary.Left);
+            var maybeRight = context.SemanticModel.GetConstantValue(binary.Right);
+
+            if (maybeLeft.HasValue ^ maybeRight.HasValue)
+            {
+                if (maybeLeft.HasValue)
+                {
+                    other = binary.Right;
+                    return TryConvertToDouble(maybeLeft.Value, out constant);
+                }
+                else
+                {
+                    other = binary.Left;
+                    return TryConvertToDouble(maybeRight.Value, out constant);
+                }
+            }
+            return false;
+        }
+
+        // 'char' needs to roundtrip {{char -> int -> double}}, can't go {{char -> double}}
+        private static bool TryConvertToDouble(object constant, out double typedConstant) =>
+            constant is char
+            ? ConversionHelper.TryConvertWith(Convert.ToInt32(constant), Convert.ToDouble, out typedConstant)
+            : ConversionHelper.TryConvertWith(constant, Convert.ToDouble, out typedConstant);
+
+        private static bool TryGetRange(ITypeSymbol typeSymbol, out double min, out double max)
```

Comment:
I think this method is big enough now to start to introduce the Range type (comply with the [ordering of the members in the class](https://github.com/SonarSource/sonar-dotnet/blob/master/docs/coding-style.md#class-members))
```suggestion
        private readonly record struct ValuesRange(double min, double max);
        
        private static ValuesRange? GetValuesRange(ITypeSymbol typeSymbol)
```

Reply:
I made it a reference type (record class), because I needed nullability for:
```csharp
TryGetRange(typeSymbolOfOther) is { } range
```